# 3章 ニューラルネットワーク

## 3.4 3層ニューラルネットワークの実装

図3-15

上図に示すような3層ニューラルネットワークを対象として、その入力から出力への処理（フォワード方向への処理）を実装する。

### 3.4.1 記号の確認

図3-16

$w_{ij}^{(k)}$: 重み<br>
$i$: 次層の$i$番目のニューロン<br>
$j$: 前層の$j$番目のニューロン<br>
$k$: 第$k$層目

バイアスは含まず、$i, j, k$は1から始まる。

### 3.4.2 各層における信号伝達の実装

入力層から第1層目の1番目のニューロンへの信号の伝達を下の図で表す。

図3-17

$b_i^{(k)}$: バイアス

$a_1^{(1)}$を数式で表すと、

$$
a_1^{(1)} = w_{11}^{(1)} x_1 + w_{12}^{(1)} x_2 + b_1^{(1)}
$$

行列の積を用いると、第1層目の「重み付き和」は次の式でまとめて表せる

$$
{\bf A}^{(1)} = {\bf X W}^{(1)} + {\bf B}^{(1)}
$$
$$
{\bf A}^{(1)} = \begin{pmatrix}
a_{1}^{(1)} & a_{2}^{(1)} & a_{3}^{(1)} \\
\end{pmatrix},
{\bf X}^{(1)} = \begin{pmatrix}
x_{1} & x_{2} \\
\end{pmatrix},
{\bf B}^{(1)} = \begin{pmatrix}
b_{1}^{(1)} & b_{2}^{(1)} & b_{3}^{(1)} \\
\end{pmatrix},\\
{\bf W}^{(1)} = \begin{pmatrix}
w_{11}^{(1)} & w_{21}^{(1)} & w_{31}^{(1)} \\
w_{12}^{(1)} & w_{22}^{(1)} & w_{32}^{(1)} \\
\end{pmatrix}
$$

NumPyの多次元配列を使って上の式を実装する。

In [3]:
import numpy as np

X = np.array([1.0, 0.5])
W1 = np.array([[0.1, 0.3, 0.5], [0.2, 0.4, 0.6]])
B1 = np.array([0.1, 0.2, 0.3])

print("W1.shape =", W1.shape)
print("X.shape =", X.shape)
print("B1.shape =", B1.shape)

A1 = np.dot(X, W1) + B1

W1.shape = (2, 3)
X.shape = (2,)
B1.shape = (3,)


続いて、第1層目の活性化関数によるプロセスを見る。この活性化関数によるプロセスは下の図のように表せる。

図3-18

$a$: 隠れ層での重み付き和（重み付き信号とバイアスの総和）<br>
$z$: 活性化関数で変換された信号
$h$: 活性化関数

活性化関数にシグモイド関数を使い、Pythonで実装すると、次のようになる。

In [4]:
def sigmoid(x):
    return 1 / (1 + np.exp(-x))

Z1 = sigmoid(A1)

print("A1 =", A1)
print("Z1 =", Z1)

A1 = [ 0.3  0.7  1.1]
Z1 = [ 0.57444252  0.66818777  0.75026011]


図3-19

続いて第1層から第2層目までの実装を行う。

In [5]:
W2 = np.array([[0.1, 0.4], [0.2, 0.5], [0.3, 0.6]])
B2 = np.array([0.1, 0.2])

print("Z1.shape =", Z1.shape)
print("W2.shape =", W2.shape)
print("B2.shape =", B2.shape)

A2 = np.dot(Z1, W2) + B2
Z2 = sigmoid(A2)

Z1.shape = (3,)
W2.shape = (3, 2)
B2.shape = (2,)


これは先程の入力層から第1層目までの実装と全く同じものになっている。

図3-20

$\sigma$: $h$とは異なる活性化関数

最後に、第2層から出力層への信号の伝達を実装する。この実装も、これまでとほとんど同じだが、最後の活性化関数だけが、これまでの隠れ層とは異なる。

In [6]:
def identity_function(x):
    return x

W3 = np.array([[0.1, 0.3], [0.2, 0.4]])
B3 = np.array([0.1, 0.2])

A3 = np.dot(Z2, W3) + B3
Y = identity_function(A3)

ここではidentity_function()という関数を定義して出力層の活性化関数として利用した。
このような、入力をそのまま出力する関数を**恒等関数**という。

```
出力層で利用する活性化関数は、解く問題の性質に応じて決める。

回帰問題…恒等関数
2クラス分類問題…シグモイド関数
多クラス分類問題…ソフトマックス関数

を使うのが一般的。詳しくは3.5節で扱う。
```

### 3.4.3 実装のまとめ

今回行った実装をまとめて記載する。ニューラルネットワークの実装の慣例として、重みだけをW1のように大文字で表記し、それ以外は小文字で表記する。

In [7]:
import numpy as np

def sigmoid(x):
    return 1 / (1 + np.exp(-x))

def identity_function(x):
    return x

def init_network():
    network = {}
    network['W1'] = np.array([[0.1, 0.3, 0.5], [0.2, 0.4, 0.6]])
    network['b1'] = np.array([0.1, 0.2, 0.3])
    network['W2'] = np.array([[0.1, 0.4], [0.2, 0.5], [0.3, 0.6]])
    network['b2'] = np.array([0.1, 0.2])
    network['W3'] = np.array([[0.1, 0.3], [0.2, 0.4]])
    network['b3'] = np.array([0.1, 0.2])
    
    return network

def forward(network, x):
    W1, W2, W3 = network['W1'], network['W2'], network['W3']
    b1, b2, b3 = network['b1'], network['b2'], network['b3']

    a1 = np.dot(x, W1) + b1
    z1 = sigmoid(a1)
    a2 = np.dot(z1, W2) + b2
    z2 = sigmoid(a2)
    a3 = np.dot(z2, W3) + b3
    y = identity_function(a3)

    return y

network = init_network()
x = np.array([1.0, 0.5])
y = forward(network, x)
print("y =", y)

y = [ 0.31682708  0.69627909]


## 3.5 出力層の設計

ニューラルネットワークは分類問題にも回帰問題にも使える。

出力層の活性化関数を変更して対応する。

ソフトマックス関数は以下のように表す。

$$
y_k = \frac{\exp(a_k)}{\sum_{i=1}^{n} \exp(a_i)}
$$

恒等関数と異なり、出力の各ニューロンが全ての入力から影響を受けている。

Pythonでの実装は次のようになる。

In [8]:
def softmax(a):
    exp_a = np.exp(a)
    sum_exp_a = np.sum(exp_a)
    y = exp_a / sum_exp_a
    return y

ソフトマックス関数の実装では指数関数の計算を行うことになり、オーバーフローが起こりうる。
例えば次のようになる。

In [9]:
a = np.array([1010, 1000, 990])
softmax(a)

/Users/a13379/.pyenv/versions/anaconda3-5.0.0/lib/python3.6/site-packages/ipykernel_launcher.py:2: RuntimeWarning: overflow encountered in exp
  
/Users/a13379/.pyenv/versions/anaconda3-5.0.0/lib/python3.6/site-packages/ipykernel_launcher.py:4: RuntimeWarning: invalid value encountered in true_divide
  after removing the cwd from sys.path.


array([ nan,  nan,  nan])

このように結果が`nan`になってしまう。

そこで次のような式変形を行い、改善する。

$$
y_k = \frac{\exp(a_k)}{\sum_{i=1}^{n} \exp(a_i)}
= \frac{C \exp(a_k)}{C \sum_{i=1}^{n} \exp(a_i)}
= \frac{\exp(a_k + \log C)}{\sum_{i=1}^{n} \exp(a_i + \log C)}
= \frac{\exp(a_k + C')}{\sum_{i=1}^{n} \exp(a_i + C')}
$$

$C'$にはどんな値を用いることも出来るが、入力信号の最大値を用いる。改善した実装は以下のようになる。

In [10]:
def softmax(a):
    c = np.max(a)
    exp_a = np.exp(a - c)
    sum_exp_a = np.sum(exp_a)
    y = exp_a / sum_exp_a
    return y

すると次のように正しく計算できる。

In [12]:
a = np.array([1010, 1000, 990])
softmax(a)

array([  9.99954600e-01,   4.53978686e-05,   2.06106005e-09])

### ソフトマックス関数の特徴

ソフトマックス関数には特徴があり、出力の総和が1になる。また、出力それぞれは0から1.0の間の実数になる。

In [15]:
a = np.array([0.3, 2.9, 4.0])
y = softmax(a)
print("y =", y)
np.sum(y)

y = [ 0.01821127  0.24519181  0.73659691]


1.0

この性質のおかげで、ソフトマックス関数の出力を「確率」として解釈することができる。

ソフトマックス関数は単調増加するため、出力の一番大きなニューロンがどれかはソフトマックス関数を適用しなくても分かる。そのため、実際の問題では出力層のソフトマックス関数は省略するのが一般的。（指数関数の計算コストを省略）

※省略するのは推論の場合のみ。学習時には省略しない。詳細は次章で取り扱う。

### 出力層のニューロンの数

多クラス分類問題では、次の図のように、分類したいクラスの数を出力層のニューロン数に設定するのが一般的である。

図3-23

## 3.6 手書き数字認識

ここから実践的な問題に取り組む。学習はすでに完了したものとして、学習済みのモデルを用いて推論処理を実装する。なお、この推論処理は、ニューラルネットワークの順方向伝播 (forward propagation) ともいう。

### 3.6.1 MNISTデータセット

データセットはMNISTという手書き数字の画像セットを利用する。

※ちなみにMNISTはエムニストと読むらしい。

データセットの構成は以下

- 0から9までの数字画像
- 訓練画像が60,000枚
- テスト画像が10,000枚
- 28x28のグレースケール (1チャンネル)
- 各ピクセルは0から255の値を取る
- それぞれの画像データに対しては対応するラベルが与えられている (7, 2, 1など)

書籍の付録で、MNISTデータセットのダウンロードから画像データのNumPy配列への変換までをサポートする便利なスクリプト mnist.py が提供されている。使い方は以下

In [16]:
import sys, os
sys.path.append(os.pardir)
from dataset.mnist import load_mnist

(x_train, t_train), (x_test, t_test) = load_mnist(flatten=True, normalize=False)

print(x_train.shape)
print(t_train.shape)
print(x_test.shape)
print(t_test.shape)

Done
Done
Done
Done
Converting train-images-idx3-ubyte.gz to NumPy Array ...
Done
Converting train-labels-idx1-ubyte.gz to NumPy Array ...
Done
Converting t10k-images-idx3-ubyte.gz to NumPy Array ...
Done
Converting t10k-labels-idx1-ubyte.gz to NumPy Array ...
Done
Creating pickle file ...
Done!
(60000, 784)
(60000,)
(10000, 784)
(10000,)


最初は時間かかるけど、２回目以降はローカルに保存したファイル (pickleファイル) の読み込みを行うだけなのですぐに処理が終了する。

オプションについては、normalize=Trueで0~255の値を0.0~1.0の値に正規化する。 flatten=Trueで入力画像を1次元配列にする。Falseなら1x28x28の3次元配列となる。one_hot_label=Trueのときは結果がone-hot表現で返される。Falseなら単に正解のラベルが返却される。

まずデータの確認をする。

In [17]:
import sys, os
sys.path.append(os.pardir)
import numpy as np
from dataset.mnist import load_mnist
from PIL import Image

def img_show(img):
    pil_img = Image.fromarray(np.uint8(img))
    pil_img.show()

(x_train, t_train), (x_test, t_test) = load_mnist(flatten=True, normalize=False)

img = x_train[0]
label = t_train[0]
print("label:", label)

print("img.shape =", img.shape)
img = img.reshape(28, 28)
print("img.shape =", img.shape)

img_show(img)

label: 5
img.shape = (784,)
img.shape = (28, 28)


flatten=Trueで読み込んでいるので、reshapeしてからimg_showに食わせている。

NumPy配列からPIL用のデータオブジェクトへの変換はImage.fromarray()で行っている。

### 3.6.2 ニューラルネットワークの推論処理

推論処理を実装する。

ネットワークは、入力層を784個、出力層10個のネットワークで構成する。

隠れ層を2つ、1つ目が50個、2つ目が100個のニューロンを持つとする。この50, 100という数字は任意の値に設定できる。

まず3つの関数 get_data, init_network, predictを定義する。

In [20]:
import sys, os
sys.path.append(os.pardir)
import numpy as np
import pickle
from dataset.mnist import load_mnist
from common.functions import sigmoid, softmax


def get_data():
    (x_train, t_train), (x_test, t_test) = load_mnist(normalize=True, flatten=True, one_hot_label=False)
    return x_test, t_test

def init_network():
    with open("sample_weight.pkl", 'rb') as f:
        network = pickle.load(f)
        
    return network

def predict(network, x):
    W1, W2, W3 = network['W1'], network['W2'], network['W3']
    b1, b2, b3 = network['b1'], network['b2'], network['b3']

    a1 = np.dot(x, W1) + b1
    z1 = sigmoid(a1)
    a2 = np.dot(z1, W2) + b2
    z2 = sigmoid(a2)
    a3 = np.dot(z2, W3) + b3
    y = softmax(a3)

    return y

sample_weight.pklには、学習済みの重みとバイアスのパラメータが辞書型の変数として保存されている。

推論してみる。

In [22]:
# MNISTデータセットを取得、ネットワークを生成
x, t = get_data()
network = init_network()

accuracy_cnt = 0
for i in range(len(x)):
    # 画像を1枚ずつ取り出し、predict関数で分類
    y = predict(network, x[i])
    p = np.argmax(y) # 最も確率の高い要素のインデックスを取得
    # 予測結果pを正解ラベルtと照合
    if (p == t[i]):
        accuracy_cnt += 1

print("Accuracy:" + str(float(accuracy_cnt) / len(x)))

Accuracy:0.9352


Accuracy:0.9352とは、93.52%正しく分類することができたことを表す。次章以降ではこの認識精度をさらに高めていく。

### 3.6.3 バッチ処理

Courseraで最初からやっていたように、データセットをまとめて行列計算することをバッチ処理という。

実装は次のようにできる。

In [28]:
x, t = get_data()
network = init_network()

batch_size = 100
accuracy_cnt = 0

for i in range(0, len(x), batch_size):
    x_batch = x[i:i+batch_size]
    y_batch = predict(network, x_batch)
    p = np.argmax(y_batch, axis=1)
    accuracy_cnt += np.sum(p == t[i:i+batch_size])

print("Accuracy:" + str(float(accuracy_cnt) / len(x)))

Accuracy:0.9352


行列計算ライブラリの最適化された実装により、バッチ処理した方が計算が速くなることが多い。

## 3.7 まとめ

本章では、ニューラルネットワークの順伝播 (forward propagation) について解説した。

# 感想

NumPyを利用することで簡単に行列計算を実装できた。

意外とボリュームがあったが、forward propagationについてはすでにCourseraで勉強していたため、それほど躓くことなく進められた。Courseraでも苦労していたbackward propagationが次章で取り扱われるが、ここを頑張って乗り越えて、しっかりと理解したい。